d2/src/main/java/com/linkedin/d2/jmx/JmxManager.java

```diff
@@ -93,6 +96,21 @@ public synchronized JmxManager registerLoadBalancerState(String name,
     return this;
   }
 
+  public synchronized JmxManager registerClusterInfo(String name, ClusterInfoItem clusterInfoItem)
+  {
+    checkReg(new ClusterInfoJmx(clusterInfoItem), name);
+
+    return this;
+  }
+
+  public synchronized JmxManager registerServiceProperties(
+      String name, LoadBalancerStateItem<ServiceProperties> serviceProperties)
+  {
+    checkReg(new ServicePropertiesJmx(serviceProperties), name);
+
+    return this;
+  }
+
```

Comment:
you will also need methods "registerClusterInfoJmxMBean" and "registerServicePropertiesJmxMBean", right?
Similar to "registerRelativeLoadBalancerStrategyJmxBean".

Reply:
They are already implemented, but under an older naming schema:

  public synchronized JmxManager registerClusterInfo

  public synchronized JmxManager registerServiceProperties
  
  I will update the commit by suffixing them with JmxMBean to make it clearer, seems like newer code is using the better naming schema.

Reply:
I thought you will need the method "registerClusterInfoJmxMBean(String, ClusterInfoJmxMBean)" (instead of passing in a ClusterInfoItem) so later the container code could call it to register a "ClusterInfoJmxMBeanIngraph", which will extend ClusterInfoJmx.

Reply:
I was trying to follow other similar methods' conventions, which all seem to accept an informational object and then create a JMX object inside the method.

Reply:
I've double-checked the container code and see what you meant, going to put up another commit for this soon.

Reply:
Done.

Reply:
IIUC (referencing the registerLoadBalancerStrategy example), there should be two methods:
1. registerClusterInfo(String, **ClusterInfoItem**) {
   checkReg(**new ClusterInfoItemJMX**(ClusterInfoItem), name)
}
, this is for pegasus, which creates the JMX object inside it.

2. registerClusterInfoJmxBean(String, **ClusterInfoJmxBean**) {
   checkReg(**ClusterInfoJmxBean**, name)
}
, this will be used by container to register a ClusterInfoJmxBeanIngraph.

 @rickzx , could you confirm this?

Reply:
I'm now having just registerClusterInfoJmxBean because I think it is easy for both pegasus and container code to instantiate a ClusterInfoJmx by doing a new . I can add the extra helper too if having two such methods is the convention.

Reply:
container won't create a ClusterInfoJmx but will create a ClusterInfoJmxBeanIngraph with a ClusterInfoItem directly, by doing a _new ClusterInfoJmxBeanIngraph(ClusterInfoItem)_.

, referencing D2JmxManagerWithSensors line 82.

Reply:
I read from the container code that xyzJmxIngraph is extending xyzJmx , so I was thinking that just providing a registration helper for xyzJmx will work for both container and pegasus. I'm referring to DegraderLoadBalancerStrategyV3JmxIngraph , which inherits DegraderLoadBalancerStrategyV3Jmx .

Reply:
After reading the container code, I think we will need both the methods for registering jmx beans directly and the methods which takes property objects and create jmx in directly from them. The former is used by some Spring-based internal apps.

So let me add the additional methods and change the callbacks to use them instead.

Reply:
I've added two more APIs to register JMX Beans for original objects directly.